src/main/scala/org/broadinstitute/dsde/workbench/sam/azure/AzureService.scala

```diff
@@ -0,0 +1,120 @@
+  // Static region in which to create all managed identities
+  private val managedIdentityRegion = Region.US_EAST
+  // Tag on the MRG to specify the Sam billing-profile id
+  private val billingProfileTag = "terra.billingProfileId"
+
+  /**
+    * Looks up a pet managed identity from the database, or creates it if one does not exist.
+    */
+  def getOrCreateUserPetManagedIdentity(user: SamUser,
+                                        request: GetOrCreatePetManagedIdentityRequest,
+                                        samRequestContext: SamRequestContext): IO[(PetManagedIdentity, Boolean)] = {
+    val id = PetManagedIdentityId(user.id, request.tenantId, request.subscriptionId, request.managedResourceGroupName)
+    for {
+      existingPetOpt <- directoryDAO.loadPetManagedIdentity(id, samRequestContext)
+      pet <- existingPetOpt match {
+        // pet exists in Sam DB - return it
+        case Some(p) => IO.pure((p, false))
+        // pet does not exist in Sam DB - create it
+        case None =>
+          for {
+            _ <- validateManagedResourceGroup(request)
+            manager <- crlService.buildMsiManager(request.tenantId, request.subscriptionId)
+            petName = toManagedIdentityNameFromUser(user)
+            context = managedIdentityContext(request, petName)
+            azureUami <- IO(manager.identities().define(petName.value)
+              .withRegion(managedIdentityRegion)
+              .withExistingResourceGroup(request.managedResourceGroupName.value)
+              .withTags(managedIdentityTags(user).asJava)
+              .create(context))
+            petToCreate = PetManagedIdentity(id, ManagedIdentityObjectId(azureUami.id()), ManagedIdentityDisplayName(azureUami.name()))
+            createdPet <- directoryDAO.createPetManagedIdentity(petToCreate, samRequestContext)
+          } yield (createdPet, true)
+      }
+    } yield pet
+  }
+
+  /**
+    * Resolves a managed resource group in Azure and returns the terra.billingProfileId tag value.
+    * This is used for access control checks.
+    */
+  def getBillingProfileId(request: GetOrCreatePetManagedIdentityRequest): IO[Option[ResourceId]] = {
+    for {
+      resourceManager <- crlService.buildResourceManager(request.tenantId, request.subscriptionId)
+      mrg <- IO(resourceManager.resourceGroups().getByName(request.managedResourceGroupName.value)).attempt
+      billingProfileOpt = mrg.toOption.flatMap(_.tags().asScala.get(billingProfileTag))
+    } yield billingProfileOpt.map(ResourceId(_))
+  }
+
+  /**
+    * Validates a managed resource group. Algorithm:
+    * 1. Resolve the MRG as the publisher
+    * 2. Get the managed app id from the MRG
+    * 3. Resolve the managed app as the publisher
+    * 4. Get the managed app "plan" id
+    * 4. Validate the plan id matches the configured value
+    */
+  private def validateManagedResourceGroup(request: GetOrCreatePetManagedIdentityRequest): IO[Unit] = {
```

Comment:
This is a separate check than user access control -- it essentially validates the managed app plan that the MRG was created from. I don't think this is spoofable but will run by AppSec.